### PR TITLE
Refactored with single dispatch generic function implementation 

### LINKED ISSUE #19912: [Good First Issue]: Refactor `torchvision` preprocessing converter into Python `singledispatch`
### Context

We strive to keep our code as clean as possible - one of the ways to do it is to use advanced Python features. One of them is `singledispatch` method, which you can read more about [here](https://docs.python.org/3/library/functools.html#functools.singledispatch).

This feature has been introduced in Python 3.8, so we were hesitant to use it, and in the end used a workaround - we download `singledispatchmethod` from [PyPI](https://pypi.org/project/singledispatchmethod/) as a dependency, for Python 3.7 only (we support a range of Python versions).

Right now we are in the process of [dropping Python 3.7 support](https://github.com/openvinotoolkit/openvino/pull/19534), which means we can use `singledispatch` freely. This task regards refactoring some code in our `torchvision` preprocessing converter to use `singledispatch` feature.

### What needs to be done?

The task regards addressing these two TODO comments:

- https://github.com/openvinotoolkit/openvino/blob/master/src/bindings/python/src/openvino/preprocess/torchvision/torchvision_preprocessing.py#L50
- https://github.com/openvinotoolkit/openvino/blob/master/src/bindings/python/src/openvino/preprocess/torchvision/torchvision_preprocessing.py#L70

The functions need to have "overloads" for different types of inputs. Right now it's being done by simple `if` conditions, but it should be done with `singledispatch`. An example implementation of `singledispatch` can be found in our [data dispatcher](https://github.com/openvinotoolkit/openvino/blob/master/src/bindings/python/src/openvino/runtime/utils/data_helpers/data_dispatcher.py).

### Example Pull Requests

An example usage of `singledispatch` in our repository can be found in https://github.com/openvinotoolkit/openvino/pull/15290, to be specific in the `data_dispatcher.py` file.

### Resources

- [What is OpenVINO?](https://github.com/openvinotoolkit/openvino#what-is-openvino-toolkit)
- [Contribution guide](https://github.com/openvinotoolkit/openvino/blob/master/CONTRIBUTING.md)
- [Blog post on contributing to OpenVINO](https://github.com/openvinotoolkit/openvino/blob/master/CONTRIBUTING.md)
- [User documentation](https://docs.openvino.ai/)
- [How to test OpenVINO Python API](https://github.com/openvinotoolkit/openvino/blob/master/src/bindings/python/docs/test_examples.md)

### Contact points

@p-wysocki 
@jiwaszki
@akuporos

### PR DESCRIPTION
**Pull request by mahimairaja**
### Details:
Added single dispatch generic function implementation

* [torchvision_preprocessing.py Line 50](https://github.com/openvinotoolkit/openvino/blob/master/src/bindings/python/src/openvino/preprocess/torchvision/torchvision_preprocessing.py#L50)

* [torchvision_preprocessing.py Line 70](https://github.com/openvinotoolkit/openvino/blob/master/src/bindings/python/src/openvino/preprocess/torchvision/torchvision_preprocessing.py#L70)

### Tickets:
Refactor torchvision preprocessing converter into Python singledispatch
Closes #19912

### CODE REVIEW

**Comment by ilya-lavrenov** [comment]
build_jenkins

**Comment by mahimairaja** [comment]
I am new to jenkins, is there any docs to build jenkins...

**Comment by mlukasze** [comment]
> I am new to jenkins, is there any docs to build jenkins...

Ilya's comment was in a fact command to run our Jenkins for you, so no, you don't have to do anything :)

**Comment by mahimairaja** [comment]
Gotcha!

**Review by p-wysocki — APPROVED** [review]
LGTM, thank you for your contribution! 

@jiwaszki do you know if we can merge this PR before https://github.com/openvinotoolkit/openvino/pull/19534 is merged? Python 3.7 does not have built-in `singledispatch`, but I see that `data_dispatcher.py` also doesn't do any magic with importing it, so I suppose this PR will work with Python 3.7, right? Until we get an answer, I'm putting on a `do_not_merge` label.

@mahimairaja could you please take a look at Python API checks? It seems that `mypy` is failing. Please keep in mind that if fixing the `mypy` issue isn't practical, [skipping the line](https://stackoverflow.com/questions/49220022/how-can-mypy-ignore-a-single-line-in-a-source-file) is an option too.

**Comment by mahimairaja** [comment]
Thank you @p-wysocki :)

**Comment by mahimairaja** [comment]
I am not sure I added the return type and the linked patches only but the build has failed...

**Comment by p-wysocki** [comment]
@mahimairaja that seems like a sporadic Azure error, I restarted the pipeline.

**Comment by p-wysocki** [comment]
```bash
./src/openvino/preprocess/torchvision/torchvision_preprocessing.py:57:32: E261 at least two spaces before inline comment
    return int(size), int(size) # type: ignore
```

File: src/bindings/python/src/openvino/preprocess/torchvision/torchvision_preprocessing.py [hunk 1]
```diff
@@ -46,15 +47,22 @@
...
-        if len(size) == 1:
-            return size[0], size[0]
-        elif len(size) == 2:
-            return size
+    raise ValueError(error_msg)
+
+
+@_setup_size.register
+def _setup_size_number(size: numbers.Number, error_msg: str) -> SequenceType[int]: 
+    return int(size), int(size) # type: ignore
```
Comment (p-wysocki) [inline src/bindings/python/src/openvino/preprocess/torchvision/torchvision_preprocessing.py:57]:
```suggestion
    return int(size), int(size)  # type: ignore
```

File: src/bindings/python/src/openvino/preprocess/torchvision/torchvision_preprocessing.py [hunk 2]
```diff
@@ -46,15 +47,22 @@
...
-    if isinstance(size, Sequence):
-        if len(size) == 1:
-            return size[0], size[0]
-        elif len(size) == 2:
-            return size
+    raise ValueError(error_msg)
+
+
+@_setup_size.register
+def _setup_size_number(size: numbers.Number, error_msg: str) -> SequenceType[int]:
```
Comment (p-wysocki) [inline src/bindings/python/src/openvino/preprocess/torchvision/torchvision_preprocessing.py:56]:
```suggestion
def _setup_size_number(size: numbers.Number, error_msg: str) -> SequenceType[int]:
```

**Comment by ilya-lavrenov** [comment]
build_jenkins

**Review by akuporos — APPROVED** [review]

**Comment by mlukasze** [comment]
> Is the PR is good to go? @p-wysocki

we are waiting for last CI checks, but from code review perspective it looks like ready to merge :)

**Comment by p-wysocki** [comment]
There were some CI failures, but nothing on your side I think. I merged master, it helped on another PR with the same issue. Let's hope CI passes now.

**Comment by p-wysocki** [comment]
@ilya-lavrenov can we please get a merge on this? I would merge it but `ci/jenkins` hasn't reported it's status yet, but all the check have passed.

**Comment by ilya-lavrenov** [comment]
build_jenkins

**Comment by ilya-lavrenov** [comment]
build_jenkins

**Comment by jiwaszki** [comment]
> LGTM, thank you for your contribution!
> 
> @jiwaszki do you know if we can merge this PR before #19534 is merged? Python 3.7 does not have built-in `singledispatch`, but I see that `data_dispatcher.py` also doesn't do any magic with importing it, so I suppose this PR will work with Python 3.7, right? Until we get an answer, I'm putting on a `do_not_merge` label.
> 

@p-wysocki yes, it will work. https://github.com/openvinotoolkit/openvino/blob/7d1b7b6becf7b75d8d7f675d5262b1182235d90f/src/bindings/python/src/openvino/runtime/utils/data_helpers/wrappers.py#L7-L12 -- only class methods are not "dispatchable" in <3.8.

I am going to remove the label.

**Review by jiwaszki — APPROVED** [review]
LGTM

**Comment by mahimairaja** [comment]
Thank you team :)

**Pull request merged**